documentation/dataplane/dash-bmv2-data-plane-app.md

```diff
@@ -0,0 +1,374 @@
+# DASH BMv2 Data Plane App HLD
+
+| Rev | Date | Author | Change Description |
+| --- | ---- | ------ | ------------------ |
+| 0.1 | 06/29/2024 | Junhua Zhai | Initial version |
+
+1. [Terminology](#1-terminology)
+2. [Background](#2-background)
+3. [Project scenario](#3-project-scenario)
+4. [Resource modeling, requirement, and SLA](#4-resource-modeling-requirement-and-sla)
+5. [System Architecture overview](#5-system-architecture-overview)
+6. [Detailed design](#6-detailed-design)
+    - [6.1. DASH metadata](#61-dash-metadata)
+    - [6.2. Basic Flow](#62-basic-flow)
+    - [6.3. Flow resimulation](#63-flow-resimulation)
+    - [6.4. HA flow](#64-ha-flow)
+    - [6.5. HA flow resimulation](#65-ha-flow-resimulation)
+    - [6.6. SAI](#66-sai)
+7. [Test Plan](#7-test-plan)
+8. [Appendix](#8-appendix)
+    
+
+## 1. Terminology
+
+| Term | Explanation |
+| --- | --- |
+| VPP | Vector Packet Processing |
+| DPAPP | Data plane Application |
+
+
+## 2. Background
+
+Referring to data path logical architecture, shown at the below figure:
+
+![dash_dpapp_overview](images/dash-bmv2-data-plane-app-overview.drawio.svg)
+
+Data plane app is another packet processing engine running on CPUs. It adds extra capacities onto DASH capable ASIC:
+- It serves the slow (exception) path of packet processing to implement complicated pipeline logic, which is hard to be done in ASIC, for example adding/updating/removing table entry inline.
+- It optionally implements some DASH-SAI APIs, which are not proper/easy to be done upon ASIC SDK.
+
+### 2.1. Scope
+
+This document only focuses on describing the design of a data-plane app example, as a proof of concept, how it cooperates with [DASH pipeline BMv2](https://github.com/sonic-net/DASH/tree/main/dash-pipeline/bmv2) to implement DASH data plane. The app will be based on [VPP](https://fd.io/).
+
+## 3. Project scenario
+
+### 3.1. Stateful packet process - flow
+- Flow Creation
+
+  In DASH pipeline, after flow keys are well extracted, packet goes to flow
+  lookup stage. It does the flow lookup. If any flow is matched, packet is
+  marked a flow-hit flag, otherwise flow-miss flag. If flow-hit, the packet
+  should bypass policy matching stages, such as ACL, routing, etc, and
+  continue doing the rest stages, like flow actions. If flow-miss, the packet
+  should do policy matching stages and then be forwarded to data-plane app.
+  The data-plane app will use dash-sai APIs to create flow in flow table, and
+  then re-inject the packet back to pipeline.
+
+- Flow Deletion
+
+  In flow lookup stage, TCP FIN/RST packet should be treated specially with
+  FLOW-DELETE hint and forwarded to data-plane app.
+- Flow Age-out
+
+  In flow lookup stage, if packet hits one flow, it will refresh flow
+  timestamp. Data-plane app periodically scans flow table and check if flow is
```

Comment:
data-plane app can help the data plane implementing the flow age-out mechanism by bridging the gap in the current data plane engine.

Reply:
"e.g. ......"